src/compiler/crystal/semantic/main_visitor.cr

```diff
@@ -1887,87 +1887,89 @@ module Crystal
     #   - Don't use the type of a branch that is unreachable (ends with return,
     #     break or with a call that is NoReturn)
     def merge_if_vars(node, cond_vars, then_vars, else_vars, before_then_vars, before_else_vars, then_unreachable, else_unreachable)
-      all_vars_names = Set(String).new
-      then_vars.each_key do |name|
-        all_vars_names << name
-      end
-      else_vars.each_key do |name|
-        all_vars_names << name
+      then_vars.each do |name, then_var|
+        else_var = else_vars[name]?
+
+        merge_if_var(name, node, cond_vars, then_var, else_var, before_then_vars, before_else_vars, then_unreachable, else_unreachable)
       end
 
-      all_vars_names.each do |name|
+      else_vars.each do |name, else_var|
         then_var = then_vars[name]?
-        else_var = else_vars[name]?
+        next if then_var
 
-        # Check whether the var didn't change at all
-        next if then_var.same?(else_var)
+        merge_if_var(name, node, cond_vars, then_var, else_var, before_then_vars, before_else_vars, then_unreachable, else_unreachable)
```

Comment:
`then_var` must be `nil` after `next`. So I'm wondering if this tiny refactor might be a bit better? It just avoids fetching the actual then_var, because the value is not used.

```suggestion
        next if then_vars.has_key?(name)

        merge_if_var(name, node, cond_vars, nil, else_var, before_then_vars, before_else_vars, then_unreachable, else_unreachable)
```

Reply:
The diff looks very weird. I don't know why... Just ignore the first 5 deleted lines in the suggestion.